### PR TITLE
Move weight_decay handling from opt to layer

### LINKED ISSUE #213: Weight_decay handling should be done in `backwarding`
When weight decay is added to a layer, `Optimizer::calculate` takes `weight_decay` term and use that to update `djdw`

I think it should be done in the `layer::backwarding` phase and `optimizer` shouldn't care about `weight_decay` term as it is not about optimizing but updating derivatives.

for example `FullyConnectedLayer::backwarding` should be like

```cpp
FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
   if(should_apply_weight_decay) {
      djdw.add_i(weight, weight_decay.lambda);
   }

   /** rest of the calculation */
}
```

### PR DESCRIPTION
**Pull request by zhoonit**
**Changes proposed in this PR:**
- remove weight_decay from `Optimizer::calculate` signature
- fix bugs that weight_decay was applying twice
- apply weight decay to `fc_layer`

please note that `conv2d_layer::backwarding` also need to handle weight
decay after this PR is merged.

Resolves #213

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

File: nntrainer/src/fc_layer.cpp [hunk 1]
```diff
@@ -200,11 +200,16 @@ Tensor FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
...
 
+  bool isL2norm = weight_decay.type == WeightDecayType::l2norm;
+
   Tensor ret = djdb.dot(weight.transpose("0:2:1"));
-  Tensor djdw = input.transpose("0:2:1").dot(djdb);
+  Tensor djdw = input.chain()
+                  .transpose("0:2:1")
+                  .dot(djdb)
+                  .applyIf(isL2norm, _LIFT(add_i), weight, weight_decay.lambda)
+                  .run();
```
Comment (jijoongmoon) [inline nntrainer/src/fc_layer.cpp:211]: Then, conv2d layer also have to handle weight decay by itself.

File: nntrainer/src/optimizer.cpp [hunk 2]
```diff
@@ -77,26 +77,16 @@ int Optimizer::initialize(TensorDim d, bool set_tensor) {
...
-                          WeightDecayParam weight_decay) {
+void Optimizer::calculate(const Tensor &djdw, const Tensor &djdb,
+                          Tensor &weight, Tensor &bias, int iteration,
+                          bool init_zero) {
   Tensor djdwAvg, djdbAvg;
-  if (weight_decay.type == WeightDecayType::l2norm) {
-    djdw = djdw.add(weight.multiply(weight_decay.lambda));
-  }
-
   float ll = popt.learning_rate;
```
Comment (jijoongmoon) [inline nntrainer/src/optimizer.cpp:84]: Good to remove this. this was duplication of weight decay.

File: nntrainer/src/fc_layer.cpp [hunk 3]
```diff
@@ -200,11 +200,16 @@ Tensor FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
     }
   }
 
+  bool isL2norm = weight_decay.type == WeightDecayType::l2norm;
```
Comment (kparichay) [inline nntrainer/src/fc_layer.cpp:203]: how about add this as a common generic "protected" function in `layer.h`, and any function which has weight can use this?

Comment (zhoonit) [reply nntrainer/src/fc_layer.cpp]: Updated, thanks for the suggestion!

Comment (kparichay) [reply nntrainer/src/fc_layer.cpp]: I am sorry about unclear comment earlier. I meant applying weight decay - operation and weights.
Independent on the type of weights, the weight decay will be applied in the same way.
IMO, it would be better to write out the operation `.applyIf(this->isWeightDecayL2Norm(), _LIFT(add_i), weight, weight_decay.lambda)` once dependent on the `weight` as an argument.

Comment (zhoonit) [reply nntrainer/src/fc_layer.cpp]: Making a new operation something like `Tensor::applyWeightDecay(weightdecay)` comes to my mind.
Do you have any better idea?
